### PR TITLE
Fix struct layouts for WGSL and SPIRV

### LINKED ISSUE #811: False positive: wrong array stride
Trying to compile the following compute shader works in 0.7 but fails in 0.8:

```glsl
#version 450
layout(local_size_x = 64) in;

struct Foo {
    uint lo;
    uint hi;
};

layout(set = 0, binding = 0) buffer FooBar {
    Foo[] foo;
};

void main() {
}
```

with:

```
wgpu error: Validation Error

Caused by:
    In Device::create_shader_module
      note: label = `foo.spv`
    Type [4] '' is invalid
    Array stride 8 is smaller than the base element size 16
```

### PR DESCRIPTION
**Pull request by kvark**
Fixes #811

### CODE REVIEW

**Review by Napokue — APPROVED** [review]
LGTM, thanks!

File: src/front/spv/mod.rs [hunk 1]
```diff
@@ -2843,17 +2875,7 @@ impl<I: Iterator<Item = u32>> Parser<I> {
...
-                None => match members.last() {
-                    Some(member) => {
-                        //TODO: is this needed?
-                        // If offsets are required, we can just put 0 here
-                        member.offset + module.types[member.ty].inner.span(&module.constants)
-                    }
-                    None => 0,
-                },
-            };
+            let offset = decor.offset.unwrap_or(0); //unexpected
```
Comment (jimblandy) [inline src/front/spv/mod.rs:2878]: I think the SPIR-V spec language that covers this is:

> Composite objects in the StorageBuffer, PhysicalStorageBuffer, Uniform, and PushConstant Storage Classes must be explicitly laid out. The following apply to all the aggregate and matrix types describing such an object, recursively through their nested types:
> – Each structure-type member must have an Offset decoration.

It seems like this wouldn't apply to structs used for local variables, say. Does that matter here?

**Review by jimblandy — APPROVED** [review]
Looks good to me.

File: src/front/spv/mod.rs [hunk 2]
```diff
@@ -191,6 +191,37 @@ impl crate::ImageDimension {
     }
 }
 
+//TODO: should this be shared with `crate::proc::wgsl::layout` logic?
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/front/spv/mod.rs:194]: Right, but this is SPV frontend, it technically doesn't need to obey to any WGSL rules.
So it's a deal between SPV-in and IR, which can be different... but it's not currently

File: src/front/spv/mod.rs (hunk 1 above)
Comment (kvark) [reply src/front/spv/mod.rs:2878]: yeah, I'm changing this to an error now. Just was lazy and didn't want to panic there.

Comment (kvark) [reply src/front/spv/mod.rs:2878]: Actually, no, we still need to support the case where offsets are not present there: for interface structs (in the `Output` storage class, for example).

**Review by jimblandy — COMMENTED** [review]
Looks good to me.

**Pull request merged**